File: src/gest/sgm_decode.py
from __future__ import annotations

import struct
from dataclasses import dataclass, field
from typing import Any, Final, Literal

from gest.sgm_constants import (
    FORMAT_VERSION,
    KIND_ARTICULATED,
    KIND_DIRECTION,
    MAGIC,
    OP_DIR_F32,
    OP_END,
    OP_FRAME,
    OP_JOINTS_F32,
    OP_STATE,
)

OpKind = Literal[
    "frame",
    "joints_f32",
    "state_index",
    "direction_f32",
]


@dataclass
class SgmDecodedOp:
    kind: OpKind
    t: float | None = None
    channel_id: int | None = None
    values: list[float] = field(default_factory=list)
    state_index: int | None = None


@dataclass
class SgmDecodedChannel:
    name: str
    kind: Literal["articulated", "direction"]
    joint_count: int | None = None
    state_count: int | None = None


@dataclass
class SgmDecoded:
    format_version: int
    fps: float
    channels: list[SgmDecodedChannel]
    ops: list[SgmDecodedOp]


class GestDecodeError(ValueError):
    """Error decoding .sgm bytecode."""


def _need(data: bytes, pos: int, n: int) -> None:
    if pos + n > len(data):
        raise GestDecodeError(f"Unexpected EOF at offset {pos} (need {n} bytes).")


def decode_sgm_bytes(data: bytes) -> SgmDecoded:
    """
    Parse .sgm v1 bytecode produced by `compile_to_bytes`.

    Returns a structural summary (channel table + flat opcode stream with frame markers).
    """
    pos = 0
    _need(data, pos, 4)
    if data[pos : pos + 4] != MAGIC:
        raise GestDecodeError(f"Bad magic: expected {MAGIC!r}, got {data[:4]!r}.")
    pos += 4

    _need(data, pos, 2 + 4 + 2)
    (fmt_ver,) = struct.unpack_from("<H", data, pos)
    pos += 2
    if fmt_ver != FORMAT_VERSION:
        raise GestDecodeError(f"Unsupported format_version {fmt_ver} (expected {FORMAT_VERSION}).")
    (fps,) = struct.unpack_from("<f", data, pos)
    pos += 4
    (n_ch,) = struct.unpack_from("<H", data, pos)
    pos += 2

    channels: list[SgmDecodedChannel] = []
    for _ in range(n_ch):
        _need(data, pos, 2)
        (kind, nlen) = struct.unpack_from("<BB", data, pos)
        pos += 2
        _need(data, pos, nlen)
        name = data[pos : pos + nlen].decode("utf-8")
        pos += nlen
        if kind == KIND_ARTICULATED:
            _need(data, pos, 4)
            (jc, sc) = struct.unpack_from("<HH", data, pos)
            pos += 4
            channels.append(
                SgmDecodedChannel(
                    name=name,
                    kind="articulated",
                    joint_count=jc,
                    state_count=sc,
                )
            )
        elif kind == KIND_DIRECTION:
            channels.append(SgmDecodedChannel(name=name, kind="direction"))
        else:
            raise GestDecodeError(f"Unknown channel kind byte {kind} for channel {name!r}.")

    ops: list[SgmDecodedOp] = []
    while pos < len(data):
        _need(data, pos, 1)
        (op,) = struct.unpack_from("<B", data, pos)
        pos += 1
        if op == OP_END:
            if pos != len(data):
                raise GestDecodeError(f"Trailing data after OP_END at offset {pos}.")
            break
        if op == OP_FRAME:
            _need(data, pos, 8)
            (t,) = struct.unpack_from("<d", data, pos)
            pos += 8
            ops.append(SgmDecodedOp(kind="frame", t=float(t)))
        elif op == OP_JOINTS_F32:
            _need(data, pos, 6)
            (cid, n_floats) = struct.unpack_from("<HI", data, pos)
            pos += 6
            _need(data, pos, 4 * n_floats)
            fmt = f"<{n_floats}f"
            vals = list(struct.unpack_from(fmt, data, pos))
            pos += 4 * n_floats
            ops.append(
                SgmDecodedOp(
                    kind="joints_f32",
                    channel_id=cid,
                    values=vals,
                )
            )
        elif op == OP_STATE:
            _need(data, pos, 4)
            (cid, si) = struct.unpack_from("<HH", data, pos)
            pos += 4
            ops.append(
                SgmDecodedOp(
                    kind="state_index",
                    channel_id=cid,
                    state_index=si,
                )
            )
        elif op == OP_DIR_F32:
            _need(data, pos, 2 + 12)
            (cid,) = struct.unpack_from("<H", data, pos)
            pos += 2
            (x, y, z) = struct.unpack_from("<fff", data, pos)
            pos += 12
            ops.append(
                SgmDecodedOp(
                    kind="direction_f32",
                    channel_id=cid,
                    values=[x, y, z],
                )
            )
        else:
            raise GestDecodeError(f"Unknown opcode 0x{op:02x} at offset {pos - 1}.")

    return SgmDecoded(format_version=fmt_ver, fps=float(fps), channels=channels, ops=ops)


def decoded_to_pose_timeline(decoded: SgmDecoded) -> list[dict[str, Any]]:
    """
    Rebuild a minimal timeline: list of { "t", "pose": { channel_name: {...} } } from decoded ops.

    Channel order in the table matches sorted names at compile time; ids index that list.
    """
    id_to_name = {i: ch.name for i, ch in enumerate(decoded.channels)}
    timeline: list[dict[str, Any]] = []
    current_t: float | None = None
    current_pose: dict[str, Any] = {}

    def flush_frame() -> None:
        nonlocal current_t, current_pose
        if current_t is not None:
            timeline.append({"t": current_t, "pose": dict(current_pose)})
        current_pose = {}

    for op in decoded.ops:
        if op.kind == "frame":
            flush_frame()
            current_t = op.t if op.t is not None else 0.0
        elif op.channel_id is None:
            continue
        else:
            name = id_to_name.get(op.channel_id)
            if name is None:
                raise GestDecodeError(f"Unknown channel_id {op.channel_id}")
            if op.kind == "joints_f32":
                entry = current_pose.setdefault(name, {})
                entry["joints"] = {"format": "raw_float32", "values": list(op.values)}
            elif op.kind == "state_index":
                entry = current_pose.setdefault(name, {})
                entry["state_index"] = op.state_index
            elif op.kind == "direction_f32":
                current_pose[name] = {"dir": list(op.values)}

    flush_frame()
    return timeline

File: demo/sgm_decode.js
/**
 * SGM v1 decoder for browser — aligned with include/sgm_v1.h and src/gest/sgm_decode.py
 */
(function (global) {
  const SGM_MAGIC = [0x53, 0x47, 0x4d, 0x01];
  const SGM_OP_FRAME = 0x30;
  const SGM_OP_JOINTS_F32 = 0x31;
  const SGM_OP_STATE = 0x32;
  const SGM_OP_DIR_F32 = 0x33;
  const SGM_OP_END = 0xff;
  const SGM_KIND_ARTICULATED = 1;
  const SGM_KIND_DIRECTION = 2;

  function decodeSgmBytes(buffer) {
    const bytes = buffer instanceof ArrayBuffer ? new Uint8Array(buffer) : buffer;
    const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
    let pos = 0;
    const need = (n) => {
      if (pos + n > bytes.length) throw new Error(`SGM EOF at ${pos}`);
    };

    need(4);
    for (let i = 0; i < 4; i++) {
      if (bytes[pos + i] !== SGM_MAGIC[i]) throw new Error("Bad SGM magic");
    }
    pos += 4;

    need(8);
    const formatVersion = dv.getUint16(pos, true); pos += 2;
    if (formatVersion !== 1) throw new Error(`Unsupported format_version ${formatVersion}`);
    const fps = dv.getFloat32(pos, true); pos += 4;
    const nCh = dv.getUint16(pos, true); pos += 2;

    const channels = [];
    for (let c = 0; c < nCh; c++) {
      need(2);
      const kind = dv.getUint8(pos);
      const nlen = dv.getUint8(pos + 1);
      pos += 2;
      need(nlen);
      const name = new TextDecoder().decode(bytes.subarray(pos, pos + nlen));
      pos += nlen;
      if (kind === SGM_KIND_ARTICULATED) {
        need(4);
        channels.push({
          name,
          kind: "articulated",
          jointCount: dv.getUint16(pos, true),
          stateCount: dv.getUint16(pos + 2, true),
        });
        pos += 4;
      } else if (kind === SGM_KIND_DIRECTION) {
        channels.push({ name, kind: "direction" });
      } else {
        throw new Error(`Unknown channel kind ${kind}`);
      }
    }

    const ops = [];
    while (pos < bytes.length) {
      need(1);
      const op = dv.getUint8(pos); pos += 1;
      if (op === SGM_OP_END) {
        if (pos !== bytes.length) throw new Error("Trailing data after OP_END");
        break;
      }
      if (op === SGM_OP_FRAME) {
        need(8);
        ops.push({ kind: "frame", t: dv.getFloat64(pos, true) });
        pos += 8;
      } else if (op === SGM_OP_JOINTS_F32) {
        need(6);
        const channelId = dv.getUint16(pos, true);
        const nFloats = dv.getUint32(pos + 2, true);
        pos += 6;
        need(nFloats * 4);
        const values = [];
        for (let i = 0; i < nFloats; i++) {
          values.push(dv.getFloat32(pos, true));
          pos += 4;
        }
        ops.push({ kind: "joints_f32", channelId, values });
      } else if (op === SGM_OP_STATE) {
        need(4);
        ops.push({
          kind: "state_index",
          channelId: dv.getUint16(pos, true),
          stateIndex: dv.getUint16(pos + 2, true),
        });
        pos += 4;
      } else if (op === SGM_OP_DIR_F32) {
        need(14);
        const channelId = dv.getUint16(pos, true); pos += 2;
        ops.push({
          kind: "direction_f32",
          channelId,
          values: [dv.getFloat32(pos, true), dv.getFloat32(pos + 4, true), dv.getFloat32(pos + 8, true)],
        });
        pos += 12;
      } else {
        throw new Error(`Unknown opcode 0x${op.toString(16)} at ${pos - 1}`);
      }
    }

    return { formatVersion, fps, channels, ops, bytes };
  }

  function decodedToTimeline(decoded) {
    const idToName = Object.fromEntries(decoded.channels.map((ch, i) => [i, ch.name]));
    const timeline = [];
    let currentT = null;
    let currentPose = {};

    const flush = () => {
      if (currentT !== null) timeline.push({ t: currentT, pose: { ...currentPose } });
      currentPose = {};
    };

    for (const op of decoded.ops) {
      if (op.kind === "frame") {
        flush();
        currentT = op.t ?? 0;
      } else if (op.channelId === undefined) {
        continue;
      } else {
        const name = idToName[op.channelId];
        if (!name) throw new Error(`Unknown channel_id ${op.channelId}`);
        if (op.kind === "joints_f32") {
          const entry = currentPose[name] ?? (currentPose[name] = {});
          entry.joints = { format: "raw_float32", values: op.values.slice() };
        } else if (op.kind === "state_index") {
          const entry = currentPose[name] ?? (currentPose[name] = {});
          entry.state_index = op.stateIndex;
        } else if (op.kind === "direction_f32") {
          currentPose[name] = { dir: op.values.slice() };
        }
      }
    }
    flush();
    return timeline;
  }

  function formatBytecodeHex(bytes, pulseIndex = 0) {
    const parts = [];
    for (let i = 0; i < bytes.length; i++) {
      const hex = bytes[i].toString(16).padStart(2, "0");
      parts.push(i === pulseIndex ? `<mark>${hex}</mark>` : hex);
      if ((i + 1) % 16 === 0) parts.push("<br>");
      else if ((i + 1) % 4 === 0) parts.push(" ");
    }
    return parts.join("");
  }

  global.GestSgm = { decodeSgmBytes, decodedToTimeline, formatBytecodeHex };
})(typeof window !== "undefined" ? window : globalThis);
